lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java

```diff
@@ -94,24 +95,38 @@ public GenericData(ClassLoader classLoader) {
   /** Return the class loader that's used (by subclasses). */
   public ClassLoader getClassLoader() { return classLoader; }
 
-  public Map<String, Conversion<?>> conversions =
+  private Map<String, Conversion<?>> conversions =
       new HashMap<String, Conversion<?>>();
 
-  public Map<Class<?>, Conversion<?>> conversionsByClass =
-      new IdentityHashMap<Class<?>, Conversion<?>>();
+  private Map<Class<?>, List<Conversion<?>>> conversionsByClass =
+      new IdentityHashMap<Class<?>, List<Conversion<?>>>();
+
+  public Iterable<Conversion<?>> getAllConversions() {
+    return conversions.values();
+  }
 
   public void addLogicalTypeConversion(Conversion<?> conversion) {
     conversions.put(conversion.getLogicalTypeName(), conversion);
-    conversionsByClass.put(conversion.getConvertedType(), conversion);
+    Class<?> type = conversion.getConvertedType();
+    if (conversionsByClass.containsKey(type)) {
+      conversionsByClass.get(type).add(conversion);
+    } else {
+      List<Conversion<?>> conversions = new ArrayList<Conversion<?>>(1);
+      conversions.add(conversion);
+      conversionsByClass.put(type, conversions);
+    }
   }
 
   @SuppressWarnings("unchecked")
   public <T> Conversion<? super T> getConversionFrom(Class<T> datumClass,
                                                      LogicalType logicalType) {
-    Conversion<?> conversion = conversionsByClass.get(datumClass);
-    if (conversion != null &&
-        conversion.getLogicalTypeName().equals(logicalType.getName())) {
-      return (Conversion<T>) conversion;
+    List<Conversion<?>> conversions = conversionsByClass.get(datumClass);
+    if (conversions != null) {
+      for (Conversion<?> conversion : conversions) {
```

Comment:
It looks like you're always doing this second lookup; why not use a map of maps?